### ISSUE TITLE
Support named virtual hosts for Web transports?

### ISSUE THREAD

**Issue by oberstet** [crossbario/crossbar#511]
Crossbar.io uses Twisted Web for Web transports in routers.

Currently, we don't support hosting multiple sites (see https://en.wikipedia.org/wiki/Virtual_hosting).

Twisted Web supports virtual hosting via a special resource type: `twisted.web.vhost.NameVirtualHost` 

See here: http://twistedmatrix.com/documents/current/web/howto/using-twistedweb.html#virtual-hosts

Technically, it would be straightforward for us to support.

However, question is: _should_ we support that? Feature creep ..

And if so, how would that look in node config?

**Comment by meejah** [comment]
Also: what (if anything?) would this change about the message-routing, etc. Without virtual-hosts, you'd (presumably) have one crossbar.io instance per domain -- *with* virtual-hosts, you could serve the different web-content from the same web-server/crossbar.io instance but would the application code (and/or configuration) then have to deal with the different domains as far as calls and subscribes?

That is, say you had a method "com.example.foo" and "io.example.foo" -- with "virtual host support" you might expect that a client visiting "example.io" wouldn't be able to call "com.example.foo" without further action. Of course, if you have them on different realms etc this *can* be made to work properly, but if you followed the examples and used ``realm1`` or similar on both you might be sad to find out clients from one domain can call methods in the other.

Perhaps some simple configuration checks would be enough to avoid these types of issues?

**Comment by oberstet** [comment]
IMO, both situations make sense: have 2 different virtual Web hosts configured with WAMP endpoints on the same or on different realms. Neither of which is an "error", and hence both configs are fine.

**Comment by meejah** [comment]
Yeah, I suppose that makes sense. For example, one might have ``example.com`` and ``mobile.example.com`` hosted as two different virtual-hosts but have one crossbar.io realm for either/both.

**Comment by mshahbazi** [comment]
I have managed to handle multiple subdomains on a single crossbar instance + nginx, though it is interesting to see virtualhosts in crossbar. 
Crossbar can handle everything nginx does for a common webserver, but virtualhosting.

**Comment by oberstet** [comment]
FWIW, here is my updated perspective: I am only mildly interested in this particular issue here, as it doesn't cut in _practice_ anyway.

Because in practice, I want to run _everything_ on TLS. And named virtual hosts don't work with that. Slightly exaggerated: "everyone should run everything on TLS or have very good reasons not to".

Therefore I am way more interested in TLS/SNI https://github.com/crossbario/crossbar/issues/1024, which is roughly similar in effect, but works with TLS ..

So IMO, we really should concentrate on #1024, and only then come back to this issue. The fact that Crossbar.io currently does not support SNI is a major omission that needs to be fixed.

**Comment by mshahbazi** [comment]
Although there might be some dirty ways to use multiple certificates on a server (e.g. using multiple IPs, SAN or something), there are a lot more reasons not to do so: performance issues, maintenance, etc.
Even though, on intranet projects, I barely use TLS (i.e. fake) for the "proceed anyway" headache..
BTW, putting the emphasis on #1024 seems reasonable, IMHO.

**Comment by oberstet** [comment]
I can understand what you refer to with performance, maintenance and usability issues. However, ideally, and in the longer run, I tend to disagree. Reasoning:

Define "intranet".

There is no such thing anymore as: outside the corp. firewall, there is evilness, and inside you are safe.

What if the attacker/eavesdropper already crossed the firewall and comes after you from "inside" (like it took over a browser's host, and runs arbitrary code from there, targeting other system _behind_ or firewall supposed to keep the evil doers "outside".

I really think the way to go is compartmentalize security in systems. Each and every part needs to protect itself as much as possible. This is a long shot of course. It'll take time to get there.
